### ISSUE TITLE
Autoencoder training

### ISSUE THREAD

**Issue by Bedinka** [Bedinka/PPI-prediction-with-Deep-Learning-Language-Model-#14]
# PPI Analysis with Deep Learning Language Model Project

## Description

Based on the Keras page: https://www.tensorflow.org/tutorials/generative/autoencoder
Created and autoencoder which were trained on the submatrix inputs to create a differently dimensioned vector for later usage. 
(X, 7, 7) --> (x, 2) 
### The changes to the file can be found [here](https://github.com/lionking0000/YangLabIntern/blob/master/PPI_W_DLLM/train_autoencoding.py).

---

### Problem Statement

Saveable trained model for creating an encdoed vector, which is based on the self-distance submatrixes. 

---

### Proposed Solution

Autoencoder training file is aviable at this name: data_autoencoding.py 
Saveable and loadable models based on the inputs:
latent_dim = how many latent dimension should the autoencoder map to 
model_path = which is model name which should be loaded  , 'dina_model_sample_%d_dim_%d_size_%d_epochs_%d_index_%d.keras' % (processed_sample, latent_dim, size, epoch , i )
processed_sample = How many pdb files should be processed 
size= size of the submatrix , if set to 7 --> (7, 7) 
SAVE = if True the modell will be saved

---

### Expected Behavior (for bugs)
[Explain what you expected to happen when following the steps above.]

---

### Actual Behavior (for bugs)
[Explain what actually happened when following the steps above.]

---

### Additional Context
[Add any additional context, screenshots, or information that may be relevant to the issue.]

---

### Related Code

`@tf.keras.utils.register_keras_serializable()
class Autoencoder(Model):

    def __init__(self, latent_dim, shape, **kwargs):
        super(Autoencoder, self).__init__(**kwargs)
        self.latent_dim = latent_dim
        self.shape = shape
        self.encoder = tf.keras.Sequential([
            tf.keras.layers.Flatten(),
            tf.keras.layers.Dense(latent_dim, activation='relu'),
        ])
        self.decoder = tf.keras.Sequential([
            tf.keras.layers.Dense(tf.math.reduce_prod(shape).numpy(), activation='sigmoid'),
            tf.keras.layers.Reshape(shape)
        ])

    def call(self, x):
        encoded = self.encoder(x)
        decoded = self.decoder(encoded)
        return decoded
    
    def get_config(self):
      config = super().get_config()
      config.update(
          {
            "latent_dim": self.latent_dim,
            "shape": self.shape
          }
      )
      return config
    
    def custom_objects():
    
      return {"Autoencoder": Autoencoder}
    
class LossHistory(callbacks.Callback):
    def __init__(self):
        super().__init__()
        self.train_losses = []
        self.val_losses = []

    def on_epoch_end(self, epoch, logs=None):
        self.train_losses.append(logs.get('loss'))
        self.val_losses.append(logs.get('val_loss'))`

`def main(latent_dim, model_name, processed_sample, size, SAVE, epoch):
  
  import feature_extraction
  interacting_prot = feature_extraction.main(processed_sample, size)
  ranges= 2000
  int = 500
  loss_history = LossHistory()
  re_a = mean_matrix_vec(interacting_prot, size)
  ca_a =ca_matrix_vec(interacting_prot, size)
  
  dist_ca_train = np.array(re_a)
  dist_ca_test =  np.array(ca_a)
  dist_ca_train = dist_ca_train.astype('float32') / 255.
  dist_ca_test = dist_ca_test.astype('float32') / 255.
  #print (dist_ca_train.shape)
  #print (dist_ca_test.shape)
  shape = dist_ca_train.shape[1:]
  if os.path.exists(model_name):
    autoencoder = tf.keras.models.load_model(model_name, custom_objects={"Autoencoder": Autoencoder} )
  else: 
    autoencoder = Autoencoder(latent_dim, shape)
    autoencoder.compile(optimizer='adam', loss=losses.MeanSquaredError())
    autoencoder.fit(dist_ca_train, dist_ca_train,
                    epochs=epoch,
                    shuffle=True,
                    validation_data=(dist_ca_test, dist_ca_test),
                    callbacks=[loss_history])

  """if SAVE:
    autoencoder = Autoencoder(latent_dim, shape)
    autoencoder.compile(optimizer='adam', loss=losses.MeanSquaredError())
    autoencoder.fit(dist_ca_train, dist_ca_train,
                    epochs=epoch,
                    shuffle=True,
                    validation_data=(dist_ca_test, dist_ca_test),
                    callbacks=[loss_history])
  else:  
    autoencoder = tf.keras.models.load_model(model_name, custom_objects={"Autoencoder": Autoencoder} ) """
  
  encoded_vectors_train = autoencoder.encoder(dist_ca_train).numpy()
  #print(encoded_vectors_train.shape)
  #print(encoded_vectors_train)
  encoded_vectors_test = autoencoder.encoder(dist_ca_test).numpy()

  #print(dist_ca_train.shape)
  #print(encoded_vectors_train.shape)

  x, y , correlation, p_value, correlation2, p_value2  = spearman ( dist_ca_train, encoded_vectors_train , ranges , int )
  x_test, y_test , correlation_test, p_value_test, correlation2_test, p_value2_test  = spearman ( dist_ca_test, encoded_vectors_test , ranges , int )
  
  print("#################", model_name)
  print(correlation, p_value, correlation2, p_value2 )
  print(correlation_test, p_value_test, correlation2_test, p_value2_test )
  plot(encoded_vectors_train, model_name)
  
  if SAVE:
    tf.saved_model.save(autoencoder, model_name)
  
  collected_data = {
     "model_name": model_name,
     "epochs" : epoch,
     "latent_dim" : latent_dim,
     "matrix_size" : size,
     "processed_sample" : processed_sample,
     "train_losses": loss_history.train_losses,
      "val_losses": loss_history.val_losses,
      "spearman_correlation": correlation ,
      "spearman_p_value": p_value,
      "spearman_correlation_dim1_2": correlation2,
      "spearman_p_value_dim1_2": p_value2
    }
  
  return encoded_vectors_train #, collected_data

if __name__ == "__main__":
    '''latent_dim = 
    model_name = 'dina_model_js.keras'
    processed_sample = 1
    size = 10'''
    #SAVE = True

    '''fout = open("summary.txt", "a")
    for i in range(10):
      model_path = 'dina_model_sample_%d_dim_%d_size_%d_index_%d.keras' % (processed_sample, latent_dim, size, i)
      print(model_path)
      fout.write(model_path+"\n")
      collected_data = main(latent_dim ,model_path, processed_sample ,size, SAVE ) #latent_dim ,model_path, processed_sample ,size
      print(collected_data)
      fout.write(str(collected_data)+"\n"+"\n")
    fout.close()'''
    main()`
---

### Priority
[Set the priority level for this issue (e.g., high, medium, low).]

---

### Due Date
[Specify a due date, if necessary.]

---
### Checklist for coding 
- [x] Modul configuration.
- [x] Run model. 
- [x] Evaluate model .
- [x] Model saving .
- [x] Model loading based on the inputs. 

### Checklist for tasks 
- [x] Reviewed the problem statement and proposed solution.
- [x] Environmental dependecy solutions.
- [x] Provided expected and actual behavior matches.
- [x] Included all relevant context and code snippets.
- [x] Commited.
